File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;

long long solution(vector<int> sequence) {
    vector<long long> v;
    long long answer = 0,mi=0,ma=0;
    for(int i=0;i<sequence.size();i++){
        if(i%2==1){
            sequence[i]*=-1;
        }
        v.push_back(sequence[i]);
        if(i>0){
            v[i]+=v[i-1];
        }
    }
    for(int i=0;i<v.size();i++){
        answer=max(answer,max(abs(v[i]-mi),abs(v[i]-ma)));
        mi=min(mi,v[i]);
        ma=max(ma,v[i]);
    }
    return answer;
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
import java.util.*;

class Solution {
    public long solution(int[] sequence) {
        List<Long> v = new ArrayList<>();
        long answer = 0, mi = 0, ma = 0;
        for (int i = 0; i < sequence.length; i++) {
            if (i % 2 == 1) {
                sequence[i] *= -1;
            }
            v.add((long) sequence[i]);
            if (i > 0) {
                v.set(i, v.get(i) + v.get(i - 1));
            }
        }
        for (int i = 0; i < v.size(); i++) {
            answer = Math.max(answer, Math.max(Math.abs(v.get(i) - mi), Math.abs(v.get(i) - ma)));
            mi = Math.min(mi, v.get(i));
            ma = Math.max(ma, v.get(i));
        }
        return answer;
    }
}